### PR TITLE
Implement workload suggestion API

### PR DESCRIPTION
**Pull request by mhutch**
Adds an API to suggest which workloads should be installed to satisfy a set of missing workload packs.

Fixes the NotImplementException in the workload manifest reader library.

### CODE REVIEW

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs [hunk 1]
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
...
+            }
+
+            foreach (var c in incompleteCandidates)
+            {
+                FindCandidates(c, incompleteCandidates);
+            }
+
+            // minimize number of unnecessary packs installed by the suggestion, then minimize number of workloads ids listed in the suggestion
+            // eventually a single pass to find the "best" would be more efficient but this this more readable and easier to tweak
+            var scoredList = completeCandidates.Select(
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:345]: nit: use better naming and extract class, extract method to reduce comments.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs [hunk 2]
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
...
+                        completeCandidates.Add(candidate);
+                    }
+                    else
+                    {
+                        incompleteCandidates.Add(candidate);
+                    }
+                }
+            }
+
+            //find all valid complete permutations by recursively exploring possible branches from a root
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:298]: nit: same here.

And a method that worth comments every 3 lines is a dangerous method to edit later. We need more test around this method. And it is better to make it a separate method, even a separate class.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs [hunk 3]
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
         /// <remarks>
         /// Used by the MSBuild workload resolver to emit actionable errors
         /// </remarks>
-        public IList<WorkloadInfo> GetWorkloadSuggestionForMissingPacks(IList<string> packId)
+        public ISet<WorkloadInfo> GetWorkloadSuggestionForMissingPacks(IList<string> packIds)
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:257]: method too long

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/IWorkloadResolver.cs [hunk 4]
```diff
@@ -9,7 +9,7 @@ public interface IWorkloadResolver
     {
         IEnumerable<WorkloadResolver.PackInfo> GetInstalledWorkloadPacksOfKind(WorkloadPackKind kind);
         IEnumerable<string> GetPacksInWorkload(string workloadId);
-        IList<WorkloadResolver.WorkloadInfo> GetWorkloadSuggestionForMissingPacks(IList<string> packId);
+        ISet<WorkloadResolver.WorkloadInfo> GetWorkloadSuggestionForMissingPacks(IList<string> packId);
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/IWorkloadResolver.cs:12]: @dsplaisted please check if the API is what you need

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs [hunk 5]
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
...
+            var incompleteCandidates = new List<WorkloadSuggestionCandidate>();
+            var completeCandidates = new HashSet<WorkloadSuggestionCandidate>();
+
+            foreach (var workload in _workloads)
+            {
+                var expanded = new HashSet<WorkloadPackId>(ExpandWorkload(workload.Value));
+                if (expanded.Any(e => requestedPacks.Contains(e)))
+                {
+                    var stillMissing = new HashSet<WorkloadPackId>(requestedPacks.Where(p => !expanded.Contains(p)));
+                    var satisfied = requestedPacks.Count - stillMissing.Count;
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:283]: not used variable?

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs [hunk 6]
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
...
+            {
+                FindCandidates(c, incompleteCandidates);
+            }
+
+            // minimize number of unnecessary packs installed by the suggestion, then minimize number of workloads ids listed in the suggestion
+            // eventually a single pass to find the "best" would be more efficient but this this more readable and easier to tweak
+            var scoredList = completeCandidates.Select(
+                c =>
+                {
+                    var installedCount = c.Packs.Count(p => installedPacks.Contains(p));
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:348]: no obvious test for the scores. This part sounds super complex. If a later change messed up the score logic, a test need to fail.

And it is complicated enough that I cannot say it really works by just reading it.

**Comment by marcpopMSFT** [comment]
Marked for 5.0.2xx.  The branch isn't open yet but retarget once it is in ~1 month.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+            /// <summary>
+            /// The workload IDs that comprise this suggestion
+            /// </summary>
+            public HashSet<WorkloadDefinitionId> Workloads { get; internal set; }
+
+            /// <summary>
+            /// How many additional (and potentionally unnecessary) packs this suggestion will result in installing
+            /// </summary>
+            public int ExtraPacks { get; internal set; }
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:180]: > How many additional (and potentionally unnecessary)

I don't understand what that means. The number could be too high due to some of the packs are unnecessary? And I need to read the comments to understand what the property does. It is a complex concept, it is fine to have a longer name.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        /// </summary>
+        public WorkloadSuggestion GetBestSuggestion() => FindBest(
+                UnsortedSuggestions,
+                (x, y) => y.ExtraPacks - x.ExtraPacks,
+                (x, y) => y.Workloads.Count - x.Workloads.Count);
+
+        /// <summary>
+        /// A partial or complete suggestion for workloads to install, annotated with which requested packs it does not satisfy
+        /// </summary>
+        class WorkloadSuggestionCandidate : IEquatable<WorkloadSuggestionCandidate>
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:134]: nit: please add explicit modifier per editorconfig

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                }
+            }
+        }
+
+        public ICollection<WorkloadSuggestion> UnsortedSuggestions { get; }
+
+        /// <summary>
+        /// Finds the best value from a list of values according to one or more custom comparators. The comparators are an ordered fallback list - the second comparator is only checked when values are equal according to the first comparator, and so on.
+        /// </summary>
+        private T FindBest<T>(IEnumerable<T> values, params Comparison<T>[] comparators)
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:100]: what's the difference between this and Max() ?

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs (hunk 8 above)
Comment (wli3) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:134]: I have problem understanding what's this class present. It is a collection of Workloads , Packs but  GatherCompleteSuggestions has a list and a set of them (collection of collection of workload?). What does the branch mean here?

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+            public bool IsComplete => UnsatisfiedPacks.Count == 0;
+
+            public bool Equals(WorkloadSuggestionCandidate? other) => other != null && Workloads.SetEquals(other.Workloads);
+
+            public override int GetHashCode()
+            {
+                int hashcode = 0;
+                foreach (var id in Workloads)
+                {
+                    hashcode ^= id.GetHashCode();
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:155]: order matters?

**Review by wli3 — CHANGES_REQUESTED** [review]
Still don't have enough test cover GatherCompleteSuggestions

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        class WorkloadSuggestionCandidate : IEquatable<WorkloadSuggestionCandidate>
+        {
+            public WorkloadSuggestionCandidate(HashSet<WorkloadDefinitionId> id, HashSet<WorkloadPackId> packs, HashSet<WorkloadPackId> unsatisfiedPacks)
+            {
+                Packs = packs;
+                UnsatisfiedPacks = unsatisfiedPacks;
+                Workloads = id;
+            }
+
+            public HashSet<WorkloadDefinitionId> Workloads { get; }
```
Comment (wli3) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:143]: possible to make it immutable?

**Comment by wli3** [comment]
I hope to have some test to cover and document GatherCompleteSuggestions. Example of "recursively explores a branching tree". What kind of candidate is not "the best"

**Comment by wli3** [comment]
@mhutch could you retarget to release/5.0.2xx? And the logic of union and sort is still hard to understand. We need to have unit tests around it. 1. to document what it does 2. prevent regression and make future change easier, considering it will be someone from SDK team to maintain this code in the long term.

**Comment by mhutch** [comment]
@wli3 i've retargeted it. could you be more specific about what you want additional tests and documentation on?

**Comment by wli3** [comment]
@mhutch sorry for being late. I missed the notification somehow. If we could write unit tests for every major methods in WorkloadSuggestionFinder.cs, it would be great. Like FindBest, FilterRedundantSuggestions,GatherCompleteSuggestions, and the constructor of WorkloadSuggestionFinder.

**Comment by mhutch** [comment]
@wli3 The constructor was already implicitly covered by `CanSuggestSimpleWorkload`, `CanSuggestTwoWorkloadsToFulfilTwoRequirements` and `CanSuggestWorkloadThatFulfillsTwoRequirements` and I didn't see a clean way to test it by itself, but I factored out a chunk of logic into `FindPartialSuggestionsAndSimpleCompleteSuggestions` and added a test for that and  `GetBestSuggestion`. I also added some more doc comments, tried to clarify naming, and moved the suggestion finder tests to their own class.

**Comment by wli3** [comment]
Sorry @mhutch I missed the notification of your changes. Please go ahead to merge it after resolving the merge conflict. I can resolve the conflict and merge it if you need.

**Review by wli3 — APPROVED** [review]

**Pull request merged**